File: leetcode/enumerated/py/49.py
# https://leetcode.com/problems/group-anagrams/

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        groups = {}
        for s in strs:
            key = [0] * 26
            for c in s:
                key[ord(c) - ord('a')] += 1

            key = tuple(key) # key should be immutable
            if key not in groups.keys():
                groups[key] = []

            groups[key].append(s)

        return groups.values()

File: leetcode/enumerated/cpp/49.cpp
// https://leetcode.com/problems/group-anagrams/
#include <vector>
#include <string>
#include <array>
#include <unordered_map>

using std::string;
using std::array;
using std::vector;
using std::unordered_map;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>> result;

        unordered_map<array<int, 26>, vector<string>> kvpair;
        for (const auto& str : strs) {
            array<int, 26> key = {0};
            for (auto ch : str) {
                key[ch - 'a']++;
            }

            kvpair[key].push_back(str);
        }

        for (const auto& key : kvpair) {
            result.push_back(key.second);
        }
        return result;
    }
};
